cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/existingdataproc/ExistingDataprocConf.java

```diff
@@ -0,0 +1,291 @@
+      }
+      labels.forEach((k, v) -> {
+        if (!LABEL_KEY_PATTERN.matcher(k).matches()) {
+          throw new IllegalArgumentException("Invalid label key " + k + ". Label key cannot be longer than 63"
+                                               + "characters and can only contain lowercase letters, numeric"
+                                               + " characters, underscores, and dashes.");
+        }
+        if (!LABEL_VAL_PATTERN.matcher(v).matches()) {
+          throw new IllegalArgumentException("Invalid label value " + v + ". Label values cannot be longer than 63"
+                                               + "characters and can only contain lowercase letters, numeric"
+                                               + " characters, underscores, and dashes.");
+        }
+      });
+    }
+    return new ExistingDataprocConf(projectId, clustername, region, accountKey, keyPair, labels);
+  }
+
+  // the UI never sends nulls, it only sends empty strings.
+  @Nullable
+  private static String getString(Map<String, String> properties, String key) {
+    String val = properties.get(key);
+    if (val != null && val.isEmpty()) {
+      return null;
+    }
+    return val;
+  }
+
+  /**
+   * Get zone from the metadata server.
+   */
+  private static String getSystemZone() {
+    try {
+      String zone = getMetadata("instance/zone");
+      // will be something like projects/<project-number>/zones/us-east1-b
+      return zone.substring(zone.lastIndexOf('/') + 1);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Unable to get the zone from the environment. "
+                                           + "Please explicitly set the zone.", e);
+    }
+  }
+
+  /**
+   * Returns the region of the given zone.
+   */
+  private static String getRegionFromZone(String zone) {
+    int idx = zone.lastIndexOf("-");
+    if (idx <= 0) {
+      throw new IllegalArgumentException("Invalid zone. Zone must be in the format of <region>-<zone-name>");
+    }
+    return zone.substring(0, idx);
+  }
+
+
+  /**
+   * Get project id from the metadata server.
+   */
+  static String getSystemProjectId() {
+    try {
+      return getMetadata("project/project-id");
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Unable to get project id from the environment. "
+                                           + "Please explicitly set the project id and account key.", e);
+    }
+  }
+
+  /**
+   * Makes a request to the metadata server that lives on the VM, as described at
+   * https://cloud.google.com/compute/docs/storing-retrieving-metadata.
+   */
+  private static String getMetadata(String resource) throws IOException {
+    URL url = new URL("http://metadata.google.internal/computeMetadata/v1/" + resource);
+    HttpURLConnection connection = null;
+    try {
+      connection = (HttpURLConnection) url.openConnection();
```

Comment:
Move the `try` after this line. The code would be cleaner:

```
HttpURLConnection connection = (HttpURLConnection) url.openConnection();
try {
  // code...
} finally {
  connection.disconnect():
}

Reply:
Done